aiida/tools/graph/age_entities.py

```diff
@@ -0,0 +1,508 @@
+                if var.aiida_cls_from is not cls_from:
+                    raise TypeError('{} has to  have {} as aiida_cls_from'.format(keyword, cls_from))
+                elif var.aiida_cls_to is not cls_to:
+                    raise TypeError('{} has to  have {} as aiida_cls_to'.format(keyword, cls_to))
+                else:
+                    return var
+            else:
+                raise TypeError('{} has to be an instance of DirectedEdgeSet'.format(keyword))
+
+        nodes = get_check_set_entity_set(nodes, 'nodes', Node)
+        groups = get_check_set_entity_set(groups, 'groups', Group)
+        nodes_nodes = get_check_set_directed_edge_set(nodes_nodes, 'nodes-nodes', Node, Node)
+        groups_nodes = get_check_set_directed_edge_set(groups_nodes, 'groups-nodes', Node, Group)
+        self._dict = dict(nodes=nodes, groups=groups, nodes_nodes=nodes_nodes, groups_nodes=groups_nodes)
+
+    @property
+    def sets(self):
+        return list(zip(*sorted(self.dict.items())))[1]
+
+    @property
+    def dict(self):
+        return self._dict
+
+    @property
+    def nodes(self):
+        return self._dict['nodes']
+
+    @property
+    def groups(self):
+        return self._dict['groups']
+
+    def __getitem__(self, key):
+        return self._dict[key]
+
+    def __setitem__(self, key, val):
+        self._dict[key] = val
+
+    def __add__(self, other):
+        new_dict = {}
+        for key in self._dict:
+            new_dict[key] = self._dict[key] + other.dict[key]
+        return Basket(**new_dict)
+
+    def __iadd__(self, other):
+        for key in self._dict:
+            self[key] += other[key]
+        return self
+
+    def __sub__(self, other):
+        new_dict = {}
+        for key in self._dict:
+            new_dict[key] = self[key] - other[key]
+        return Basket(**new_dict)
+
+    def __isub__(self, other):
+        for key in other.dict:
+            self[key] -= other[key]
+        return self
+
+    def __len__(self):
+        return sum([len(s) for s in self.sets])
+
+    def __eq__(self, other):
+        for key in self._dict:
+            if self[key] != other[key]:
+                return False
+        return True
+
+    def __ne__(self, other):
+        return not self == other
+
+    def __repr__(self):
+        ret_str = ''
+        for key, val in self._dict.items():
+            ret_str += '  ' + key + ': '
+            ret_str += str(val) + '\n'
+        return ret_str
+
+    def empty(self):
+        """
+        Empty every subset from its content
+        """
+        for set_ in self._dict.values():
+            set_.empty()
+
+    def get_template(self):
+        new_dict = dict()
+        for key, val in self._dict.items():
+            new_dict[key] = val.get_template()
+        return Basket(**new_dict)
+
+    def copy(self):
+        new_dict = dict()
+        for key, val in self._dict.items():
+            new_dict[key] = val.copy()
```

Comment:
can these values be nested? i.e. do we need a `deepcopy` potentially?

Reply:
Just for the record I'll leave here the conclusions of our discussion on this:

All `_dict` elements of a `basket` are currently of type `EntitySet` or `EdgeSet`, which have a `copy` method but not a `deepcopy` one. We might want to change this later but I think for now it should remain as is. This `copy` method has been implemented in the parent `AbstractSetContainer` class and removed from the children.